### ISSUE TITLE
AMOEBA+ migrating to the latest OpenMM

### ISSUE THREAD

**Issue by leucinw** [openmm/openmm#2375]
This issue is a discussion of AMOEBA+ implementation in OpenMM and how to migrate this part of code to the latest OpenMM. 

AMOEBA+ model was implemented based on the previous AMOEBA plugin code in OpenMM. Here are new features comparing to AMOEBA:  

1.  vdW term: W-H combining rule for epsilon was added. https://github.com/leucinw/AMOEBAplusCFlux/blob/master/OpenMM/plugins/amoeba/openmmapi/src/AmoebaVdwForce.cpp#L182-L189

2. charge transfer (CT) term: a non-bonded interaction with user defined exclusion rules (1-2, 1-3, etc). This is implemented as an individual energy term in the same manner as AMOEBA vdW term. Except that CT does not use induction factor for hydrogen (i.e. induction factor = 1.0). Openmmapi (https://github.com/leucinw/AMOEBAplusCFlux/blob/master/OpenMM/plugins/amoeba/openmmapi/src/AmoebaCTForce.cpp) and cuda platform (https://github.com/leucinw/AMOEBAplusCFlux/blob/master/OpenMM/plugins/amoeba/platforms/cuda/src/AmoebaCudaKernels.cpp#L2979)

3. polarization: AMOEBA+ uses a new damping scheme for the direct induction, while keeps the Thole damping for the mutual part (which AMOEBA model uses). The direct damping functions used with Spherical multipoles are implemented (https://github.com/leucinw/AMOEBAplusCFlux/blob/master/OpenMM/plugins/amoeba/platforms/cuda/src/kernels/pmeMultipoleElectrostatics.cu#L159) .
Cartesian multipole based damping functions were implemented in CPU (https://github.com/leucinw/AMOEBAplusCFlux/blob/master/Tinker/epolar1.f#L306)

4. charge penetration (CP): we use slightly different damping functions than HIPPO model. Again, damping functions for Cartesian multipoles are here: https://github.com/leucinw/AMOEBAplusCFlux/blob/master/Tinker/empole1.f#L375
CP was implemented with Cartesian multipoles based code as a correction to the force and energy (https://github.com/leucinw/AMOEBAplusCFlux/blob/master/OpenMM/plugins/amoeba/platforms/cuda/src/kernels/pmeMultipoleElectrostatics.cu#L455). This implementation roughly slows down ~20% because we calculated nearly "twice" of the multipole-multipole interaction, one is normal electrostatics and the other is the CP correction. 

5. charge flux (CF):  First I looped over bond (https://github.com/leucinw/AMOEBAplusCFlux/blob/master/OpenMM/plugins/amoeba/platforms/cuda/src/kernels/amoebaCFluxAndForce.cu#L3) and angle (https://github.com/leucinw/AMOEBAplusCFlux/blob/master/OpenMM/plugins/amoeba/platforms/cuda/src/kernels/amoebaCFluxAndForce.cu#L33) to accumulate charge fluxes. Then I added the accumulated charges to the force field defined charges to get the updated charges. In the force and energy kernels, I also accumulated potential on each atom(https://github.com/leucinw/AMOEBAplusCFlux/blob/master/OpenMM/plugins/amoeba/platforms/cuda/src/kernels/pmeMultipoleElectrostatics.cu#L445 , https://github.com/leucinw/AMOEBAplusCFlux/blob/master/OpenMM/plugins/amoeba/platforms/cuda/src/kernels/pmeMultipoleElectrostatics.cu#L802 , https://github.com/leucinw/AMOEBAplusCFlux/blob/master/OpenMM/plugins/amoeba/platforms/cuda/src/kernels/pmeMultipoleElectrostatics.cu#L831 , https://github.com/leucinw/AMOEBAplusCFlux/blob/master/OpenMM/plugins/amoeba/platforms/cuda/src/kernels/multipolePme.cu#L864 , https://github.com/leucinw/AMOEBAplusCFlux/blob/master/OpenMM/plugins/amoeba/platforms/cuda/src/kernels/multipolePme.cu#L950). After the force and energy calculation, a chain rule term contribute to force is calculated. ( https://github.com/leucinw/AMOEBAplusCFlux/blob/master/OpenMM/plugins/amoeba/platforms/cuda/src/kernels/amoebaCFluxAndForce.cu#L84-L221)

*6. ind&inp induced dipole: currently in AMOEBA we have two sets of induced dipoles. In AMOEBA+ CPU code, I have merged them and use only one set of dipole. We used two sets of scaling factors (polar-1x-intra and polar-1x-inter). (https://github.com/leucinw/AMOEBAplusCFlux/tree/master/Tinker). 

 I am expecting this mergence can speed up the computations.

**Comment by peastman** [comment]
Thanks!  That's a lot of pieces, so let's maybe take them a few at a time.  I'm expecting some will be very simple (just minor changes to existing code), while others will be much bigger changes, possibly requiring completely new force classes.

Are there papers describing the various interactions?  That would be very helpful for understanding what needs to be implemented.

Here are some thoughts on a few of the items.

1. This is likely to be the simplest one.  AmoebaVdwForce already supports several combining rules, so adding one more should be easy.  If you want to create a pull request for this one as a way of getting your feet wet, that would be a good place to start.

2. Can this be done with a CustomNonbondedForce?  If so, that's definitely the easy way to do it.

3, 4, and 5. These all involve changes to the multipole interactions.  The changes for 3 and 4 will probably not be too bad, but 5 could require pretty drastic changes.  I'm therefore wondering if it would be best to make a completely new force for AMOEBA+ multipole interactions, the same way we did for HIPPO?

6. Is this one necessary for correctness, or is it just an optimization?  From my experience implementing HIPPO, the performance effect of merging the dipoles is modest but certainly not negligible.  The biggest impact is on reciprocal space.  We can spread the charges onto a real grid rather than a complex one, and then use a real-to-complex FFT instead of a complex-to-complex one.

**Comment by leucinw** [comment]
These interactions are described in AMOEBA+ paper (https://pubs.acs.org/doi/10.1021/acs.jctc.9b00261) except for the charge flux term. The two damping scheme polarization is described in detail here: https://pubs.acs.org/doi/abs/10.1021/acs.jctc.7b00225 , and charge penetration is detailed in https://pubs.rsc.org/en/content/articlehtml/2016/cp/c6cp06017j

1. I will try to make a pull request.

2. Yes the CT interactions can be done with CustomNonbondedForce. See https://github.com/leucinw/AMOEBAplusCFlux/blob/master/OpenMM/plugins/amoeba/platforms/cuda/src/AmoebaCudaKernels.cpp#L3052 and https://github.com/leucinw/AMOEBAplusCFlux/blob/master/OpenMM/plugins/amoeba/platforms/cuda/src/kernels/amoebaCTForce.cu

5. In my implementation, charge flux is part of multipole interactions. So this changes the AmoebaMultipole interactions. If we want backward compatibility to run AMOEBA, we should definitely do some #ifdef detectives. I think either making a new force for AMOEBA+ multipole interactions or modifying on Amoeba code should work for our purpose. It will be more clear and easier to maintain if we make a complete new force ? The efficiency will probably be similar between the two approaches ?

6. The charge flux term we added needs to set the scaling factors (1-2, and 1-3) to 0. So we think it is necessary to make atom-based scaling factors. So this is more for correctness.

**Comment by peastman** [comment]
> Yes the CT interactions can be done with CustomNonbondedForce

That's a handwritten CUDA implementation.  I meant the [CustomNonbondedForce](http://docs.openmm.org/latest/api-c++/generated/OpenMM.CustomNonbondedForce.html#_CPPv2N6OpenMM20CustomNonbondedForceE) class.

> So this is more for correctness.

That's another argument for having a separate force class for AMOEBA+ multipoles then.  It sounds like that's probably the best approach.

**Comment by leucinw** [comment]
> That's a handwritten CUDA implementation. I meant the [CustomNonbondedForce](http://docs.openmm.org/latest/api-c++/generated/OpenMM.CustomNonbondedForce.html#_CPPv2N6OpenMM20CustomNonbondedForceE) class.

Yes I think we can use CustomNonbondedForce class, using cutoff and switching function. 

> That's another argument for having a separate force class for AMOEBA+ multipoles then. It sounds like that's probably the best approach.

I agree.

**Comment by peastman** [comment]
Note that even if it's a separate Force class, it can still use a lot of the same CUDA kernels as the standard one, possibly with just a few ifdefs added to distinguish the cases.  No need to duplicate all that code.

**Comment by peastman** [comment]
Now that the AmoebaVdwForce changes are done, we should start discussing how to implement the rest.  Since the plan is to create a new force class (AmoebaPlusNonbondedForce?), we should consider ways of doing it.  I see two main possibilities.

1. Start by cloning AmoebaMultipleForce and adapt it.  This would likely be the easiest way, since your existing implementation in Tinker-OpenMM is based on AmoebaMultipoleForce (though a heavily modified version).

2. Start by cloning HippoNonbondedForce and adapt it.  This might require a bit more adapting, but in some ways it's actually closer to AMOEBA+.  For example, it already has the d and p dipoles unified.  And although the damping functions in HIPPO and AMOEBA+ are different, they both come into the code in basically the same way.

Basing it on HippoNonbondedForce would have two significant advantages.  First, the code is *much* cleaner, so you would ultimately end up with a simpler implementation.  And second, it also is faster.  It does various things in more modern ways that allow it to take advantage of newer optimizations.

**Comment by jchodera** [comment]
From the perspective of someone who wants to experiment with new functional forms, the API is much more useful for this purpose if each interaction component is separate, rather than consolidating a great number of different kinds of interactions into a single force (like Hippo, or even the AMOEBA nonbonded force that rolls polarizable dipoles and fixed multiples into a single Force). I realize this makes it difficult to also encapsulate the code within a force in a modular way because it introduces interforce dependencies, but it very much helps experimentation if components can be mixed and matched.

If the difficulty for OpenMM is that some of these forces are essentially incompatible (eg NonbondedForce monopoles and Amoeba multiples don't share information), is there an alternative concept that could be useful, like a related compatible set of force terms that share information with a more granular way of specifying the force components of interest?

**Comment by peastman** [comment]
In the case of HIPPO, the thing that necessitated combining them is that they share parameters.  For example, the electrostatic force and the Pauli repulsion force both use exactly the same multipole moments.  Because they're the same, you 1) only need to load them once, and 2) only need to do a bunch of calculations once instead of twice.  We could have separated them, but then it would have been a lot slower.

In the case of AMOEBA+, the vdw part is already a separate class.  Of the rest, charge transfer is the only part that's clearly separable (and that can already be done with a custom force).

**Comment by leucinw** [comment]
> 2. Start by cloning HippoNonbondedForce and adapt it.  This might require a bit more adapting, but in some ways it's actually closer to AMOEBA+.  For example, it already has the d and p dipoles unified.  And although the damping functions in HIPPO and AMOEBA+ are different, they both come into the code in basically the same way.
> 
> Basing it on HippoNonbondedForce would have two significant advantages. First, the code is _much_ cleaner, so you would ultimately end up with a simpler implementation. And second, it also is faster. It does various things in more modern ways that allow it to take advantage of newer optimizations.

I agree we do AMOEBA+ (AmoebaPlusNonBondedForce) with the latter approach if this will gain faster computation and cleaner code, which will be easy to implement new features in the future. 

This part is probably simpler than or similar to Hippo(?), considering AMOEBA+ will have __charge penetration__, __dual-damping-polarization__, and __charge flux__, while Hippo also has __pauli-repulsion__ and __dispersion__.

**Comment by peastman** [comment]
The charge penetration should be easy to convert, since it's just a different functional form.  And you can easily strip out the Pauli repulsion and dispersion parts.

Charge flux is likely to be the most complicated part, because it's fundamentally different from anything in AMOEBA or HIPPO.

**Comment by peastman** [comment]
Just wanted to check in on this.  Have you started work on it?  Let me know if you have any questions.

**Comment by leucinw** [comment]
Not yet while I am also migrating the current CPU code to the latest Tinker. Do you have time working on OpenMM?

**Comment by peastman** [comment]
That's fine, I just wanted to check in.  I'm happy to help with advice and answering questions, but you'll have to do most of the coding.

**Comment by peastman** [comment]
@leucinw has anything further happened with this?
